File: app/src/main/java/com/amplifyframework/datastore/sample/LocalDataStoreInteractor.java
package com.amplifyframework.datastore.sample;

import android.util.Pair;

import com.amplifyframework.datastore.generated.model.Post;
import com.amplifyframework.rx.RxDataStoreCategoryBehavior;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

final class LocalDataStoreInteractor implements LocalPresentation.PostInteractor {
    private final RxDataStoreCategoryBehavior dataStore;

    LocalDataStoreInteractor(RxDataStoreCategoryBehavior dataStore) {
        this.dataStore = dataStore;
    }

    @Override
    public Single<Post> createRandom() {
        Post post = Posts.random("Local");
        return dataStore.save(post).toSingleDefault(post);
    }

    @Override
    public Single<List<Post>> updateAll(List<Post> posts) {
        return Observable.fromIterable(posts)
            .map(Posts::update)
            .flatMapSingle(updatedPost -> dataStore.save(updatedPost).toSingleDefault(updatedPost))
            .toList();
    }

    @Override
    public Single<List<Post>> list() {
        return dataStore.query(Post.class).toList();
    }

    @Override
    public Completable deleteAll(List<Post> posts) {
        return Observable.fromIterable(posts)
            .flatMapCompletable(dataStore::delete);
    }

    @Override
    public Observable<Pair<Post, Modification>> subscribe() {
        return dataStore.observe(Post.class)
            .map(change -> {
                Modification modification = Modification.from(change.type());
                return Pair.create(change.item(), modification);
            });
    }
}

File: app/src/main/java/com/amplifyframework/datastore/sample/LocalDataStoreInteractor.kt
package com.amplifyframework.datastore.sample

import android.util.Pair

import com.amplifyframework.datastore.generated.model.Post
import com.amplifyframework.datastore.sample.LocalPresentation.PostInteractor
import com.amplifyframework.rx.RxAuthCategoryBehavior
import com.amplifyframework.rx.RxDataStoreCategoryBehavior

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import java.lang.RuntimeException

internal class LocalDataStoreInteractor(
        private val auth: RxAuthCategoryBehavior,
        private val dataStore: RxDataStoreCategoryBehavior) : PostInteractor {
    override fun createRandom(): Single<Post> {
        val post = Posts.random("Local")
        return dataStore.save(post).toSingleDefault(post)
    }

    override fun updateAll(posts: List<Post>): Single<List<Post>> {
        return Observable.fromIterable(posts)
            .map { Posts.update(it) }
            .flatMapSingle { dataStore.save(it).toSingleDefault(it) }
            .toList()
    }

    override fun list(): Single<List<Post>> = dataStore.query(Post::class.java).toList()

    override fun deleteAll(posts: List<Post>): Completable =
        Observable.fromIterable(posts).flatMapCompletable { dataStore.delete(it) }

    override fun subscribe(): Observable<Pair<Post, Modification>> {
        return dataStore.observe(Post::class.java)
            .map { Pair.create(it.item(), Modification.from(it.type())) }
    }

    override fun signIn(username: String, password: String): Completable {
        return auth.signIn(username, password)
            .flatMapCompletable {
                if (it.isSignInComplete) {
                    Completable.complete()
                } else {
                    Completable.error(RuntimeException("Sign-in failure! Next step = ${it.nextStep}"))
                }
            }
    }
}
